tfprotov5/tf5server/server.go

```diff
@@ -130,61 +184,170 @@ func (s *server) stoppableContext(ctx context.Context) context.Context {
 	return stoppable
 }
 
+// loggingContext returns a context that wraps `ctx` and has
+// terraform-plugin-log loggers injected.
+func (s *server) loggingContext(ctx context.Context) context.Context {
+	// generate a request ID
+	reqID, err := uuid.GenerateUUID()
+	if err != nil {
+		reqID = "unable to assign request ID: " + err.Error()
+	}
+
+	// set up the logger SDK loggers are derived from
+	ctx = tfsdklog.NewRootSDKLogger(ctx, append(tfsdklog.Options{
+		tfsdklog.WithLevelFromEnv("TF", "LOG", "SDK"),
+	}, s.tflogSDKOpts...)...)
+	ctx = tfsdklog.With(ctx, "tf_req_id", reqID)
+	ctx = tfsdklog.With(ctx, "tf_provider_addr", s.name)
+
+	// set up our protocol-level subsystem logger
+	ctx = tfsdklog.NewSubsystem(ctx, tflogSubsystemName, append(tfsdklog.Options{
+		tfsdklog.WithLevelFromEnv("TF", "LOG", "SDK", "PROTO"),
```

Comment:
Similarly with this environment variable.

Reply:
Conversely, are people going to know to look for line 218 to find this environment variable name while looking through this code? It's also used in line 264.

Reply:
Who is "people", here? I think provider devs/practitioners would know about the environment variable from the website, not from reading the code, whether it's a constant or inlined.

Reply:
An example person is myself, a maintainer that will most likely not look at this code for months, then try to reassemble context on what environment variables are used or are implemented where. In this case searching for `TF_LOG_SDK_PROTO` will not yield anything in the code.

Reply:
Yeah, the search argument is good. I've updated it to use the full environment variable. Can you help me understand why you'd be looking for a list of environment variables used? My usual experience with environment variables is I'm looking for a specific variable, and find it via grep. If that grep leads me to a constants file, I then need to grep for the constant to find where it's used, and that's harder, because it can take multiple forms (`MyConstant` and `mypkg.MyConstant`, because `MyConstant` may be reused as an identifier in a different package).

Reply:
This is related to https://github.com/hashicorp/terraform-plugin-log/issues/12, which explicitly notes:

> Proper log handling in Terraform is complex! ... and the plethora of TF_LOG* environment variables that are available to control logging

Having a single, easy to find place with those environment variables in the code (and which ones are explicitly handled in _this_ code) makes that discovery easier. Honestly, it doesn't matter too much to me how it is documented in the code, I just feel it can be more difficult to know how this particular thing applies to the overall system or code when it is buried in the middle of logic.

I guess this boils down to a heuristic I mentally try to apply to my own code. What outside things might affect or be affected by this code, potentially unexpectedly? Those tend to be treated more special in terms of documentation or coding patterns. Environment variables tends to meet that bar for me because it is a side-effect or logic branch due to the execution environment. (The log keys present in output as noted in the other thread also do for their potential effects on consumers if they get changed.)

> If that grep leads me to a constants file, I then need to grep for the constant to find where it's used, and that's harder, because it can take multiple forms (MyConstant and mypkg.MyConstant, because MyConstant may be reused as an identifier in a different package).

Language servers can infer proper references versus generic text tooling like grep. Decent naming can also help, as I would hope this would never be named something generic like that. Something like `envTfLogSdkProto` doesn't seem like it would be too common elsewhere and cues to the fact it is something environment related to Terraform, logging, the SDK, and protocol.

---

Regardless of constant or not, my main point here is that I personally don't want to continue a game of "find the environment variable and figure out what it does or how it is affected by other environment variables" every time a bug report or feature request comes in for Terraform's logging.

I have found that a quick way to do that is a constant definition that explains its purpose and expected behaviors:

```go
// Environment variables
const (
  // Logging level for Terraform protocol logging.
  // Includes protocol call information and handoffs to plugins.
  // Does not include request or response data for those calls, see TF_LOG_SDK_PROTO_DATA.
  // Infers logging level from SDK loggers and above.
  // Valid values match other loggers, except JSON.
  envTfLogSdkProto = "TF_LOG_SDK_PROTO"
)
```

Another quick way is code comments where it is used with the same explanations, or even comments at the package level. A longer form way to do that is separate hard copy, such as intended by https://github.com/hashicorp/terraform-plugin-log/issues/12, although that can be trickier to keep up to date and does not necessarily help while you're looking at the code within the codebase itself.

Reply:
I think the disagreement here is less about whether we need to document environment variables and more about whether constants are necessary to document environment variables and where the appropriate place(s) to document environment variables is.